Hydro devel

PaulBouchier opened the pull request:
Introduced protocol version in 2nd byte of messages. 

@mikeferguson, please have a look at the change in node_handle.h to pre-enable the NEXT protocol change to be detected by rosserial_python: if client detects a protocol mismatch when configured_ is false, it sends a requestTimeSync() message. This will be ignored in the case of protocol mismatch, and seems it should be harmless if the protocol mismatch was due to data corruption and in fact there is not a protocol mismatch. 

@MikePurvis - thanks for suggestion to send a "trial message" using old protocol. I didn't go that route because the protocol is stateless at present - that would have made it stateful inasmuch as each side would have had to go from an uninitialized state to an initialized state  in sync with the other and I'm worried about a bunch of corner cases popping out of that where one side is initialized and the other not, and recovery from that. Seems like a can of worms. Instead I opted for the best messages I could emit to suggest that there might be a protocol mismatch in the case of non-response.

PaulBouchier committed:
SerialClient.py:
- Changed 2nd sync byte to a protocol version 0xfe
- Changed message header state machine to check protocol version
- Changed 2nd sync mismatch message to express protocol version mismatch
- Created a "synced" state, initially false. Sync timeout message if
protocol has never synced suggests broken link or version mismatch, whereas
if protocol has synced it uses old timeout-restarting message

node_handle.h:
- Changed 2nd sync byte to a protocol version 0xfe
- Changed message header state machine to check protocol version
- Changed handling of protocol byte mismatch to cause it to send a
timeSyncRequest back to rosserial_python, which triggers rosserial_python
to detect the protocol mismatch & emit an error message. This won't do
anything for hydro-rosserial_python + groovy-rosserial_client, because
groovy rosserial_client doesn't have this logic, but next time we change
the protocol it will kick in & trigger a meaningful error message from
rosserial_python

Fixed committer ID

PaulBouchier committed:
Change tabs to spaces in previous commit

PaulBouchier commented:
BTW, this commit incorporates tongtybj's pull request to groovy, which we decided to defer to hydro.

mikeferguson commented on rosserial_python/src/rosserial_python/SerialClient.py:
```diff
@@ -319,6 +320,12 @@ def __init__(self, port=None, baud=57600, timeout=5.0):
         time.sleep(0.1)           # Wait for ready (patch for Uno)
 
+        # hydro introduces protocol ver2 which must match node_handle.h
+	# The protocol version is sent as the 2nd sync byte emitted by each end
```
You have a tab here -- there should be no tabs in our python code (as per ROS coding standard).

mikeferguson commented on rosserial_python/src/rosserial_python/SerialClient.py:
```diff
@@ -348,35 +355,32 @@ def requestTopics(self):
-                rospy.logerr("Lost sync with device, restarting...")
+                if (self.synced == True):
+                    rospy.logerr("Lost sync with device, restarting...")
+		else:
```
Tabs here as well.

mikeferguson commented on rosserial_python/src/rosserial_python/SerialClient.py:
```diff
@@ -348,35 +355,32 @@ def requestTopics(self):
+            if ( flag[1] != self.protocol_ver):
+                self.sendDiagnostics(diagnostic_msgs.msg.DiagnosticStatus.ERROR, "Mismatched protocol version in packet: lost sync or rosserial_python is from different ros release than the rosserial client")
+                rospy.logerr("Mismatched protocol version in packet: lost sync or rosserial_python is from different ros release than the rosserial client")
+                rospy.loginfo("Protocol version was %s, expected %s " % (flag[1], self.protocol_ver))
```
It might helpful to have a string output here, for instance, if it said "Protocol version was Rev. 0 (rosserial 0.4 and earlier), expected Rev. 1 (rosserial 0.5+)". This could be accomplished by having a little dictionary of protocol characters mapping to pretty print outputs describing them -- I think this would help a lot of users avoid confusion.

mikeferguson commented:
I added a couple of comments in the code. Generally looks OK to me -- I presume you've tested on embedded linux, how about Arduino?

PaulBouchier committed:
- Remove some more tabs that vi inserted.
- Added dictionary to improve protocol version message

PaulBouchier commented:
Fixed the tabs & changed my .exrc so it won't happen again. Added a dictionary as requested. All my testing has been done on Arduino with a 328 processor. I don't have an Atmel 168 handy, and my experience is it can't do much anyway - runs out of memory too fast. I re-checked and rosserial_embeddedlinux still runs fine, so I merged. I'll start the release procedure.

Since I merged tongtybj's work, I guess I should close his pull request?

mikeferguson commented:
If it works on a 328, should be fine on any of the regular arduinos, so I wouldn't worry about that.

And  yeah, close the pull request with a note to this one.

mikepurvis commented:
Just looking at this now—not sure I understand how it works. It looks like it only ever sends a new-style sync message in the requestTopics() call. How does the warning message get triggered?

PaulBouchier commented:
The warning message doesn't get triggered with a groovy Arduino, because groovy Arduino throws away the hydro packet. However, in the future, if we have another protocol change, say in i-ros, then a hydro-arduino will respond to a protocol mismatch by sending a request timesync, at which point i-ros rosserial_python will detect the protocol mismatch and issue the warning. I am putting in architecture to better support future protocol changes.

For groovy arduino, the change is to the timeout message (hydro rosserial_python times out when groovy Arduino doesn't respond). The timeout message suggests a couple of possible causes: possible link problem or mismatched SW versions.
